### PR TITLE
Support for primitive types in ortmodule

### PR DESCRIPTION
**Pull request by baijumeswani**
**Description**: With this change, users can input ```int```s, ```bool```s and ```float```s to ```ORTModule``` forward method.

The approach in this pull request is to wrap primitive types inside tensors before being executed (by the exporter as well as by the backend).

An overly simplistic example of a model that uses primitive types is:

```py
class PrimitiveTypesInputNet(torch.nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(PrimitiveTypesInputNet, self).__init__()

        self.fc1 = torch.nn.Linear(input_size, hidden_size)
        self.relu = torch.nn.ReLU()
        self.fc2 = torch.nn.Linear(hidden_size, num_classes)

    def forward(self, input1, bool_argument, int_argument, float_argument):
        input1 = input1 + int_argument + float_argument
        if bool_argument:
            out = self.fc1(input1)
            out = self.relu(out)
            out = self.fc2(out)
        else:
            out = self.fc1(input1)
            out = self.fc2(out)
            out = self.relu(out)
        return out

bool_argument = True
int_argument = 123
float_argument = 1.234

device = 'cuda'
N, D_in, H, D_out = 32, 784, 500, 10
model = PrimitiveTypesInputNet(D_in, H, D_out).to(device)
model = ORTModule(model)

out = model(input1, bool_argument, int_argument, float_argument)
```

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 1]
```diff
@@ -9,6 +9,22 @@
 import torch
 import warnings
 
+class _Primitive(object):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:12]:
```suggestion
class _PrimitiveType(object):
```

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 2]
```diff
@@ -9,6 +9,22 @@
...
+    def is_primitive(value):
+        return type(value) in _Primitive._primitive_types
+
+    @staticmethod
+    def get_tensor(value, device):
+        return torch.tensor(value, device=device)
+
+    @staticmethod
+    def get_primitive_dtype(value):
+        # If the data is a boolean primitive, save the value of the boolean in dtype.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:24]:
```suggestion
        # If `value` is a boolean, save the value of the boolean in dtype.
```

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 3]
```diff
@@ -40,11 +56,20 @@ def __repr__(self) -> str:
...
+    def flatten(self, args, kwargs, device):
         '''Flatten args and kwargs in a single tuple of tensors with strict ordering'''
 
-        ret = list(args)
+        ret = [None]*len(args)
+        for idx, arg in enumerate(args):
+            if _Primitive.is_primitive(arg):
+                ret[idx] = _Primitive.get_tensor(arg, device)
+            else:
+                ret[idx] = arg
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:67]: List comprehension performs better than regular loop with if-else. This perf gain is important as flattening happens for all steps

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 4]
```diff
@@ -40,11 +56,20 @@ def __repr__(self) -> str:
...
         ret += [kwargs[name] for name in self.names if name in kwargs]
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:68]: isnt it possible to have int,bool or float in kwargs as in `**{'baiju_is_cool': True}`

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 5]
```diff
@@ -40,11 +56,20 @@ def __repr__(self) -> str:
...
+        for idx in range(len(args), len(ret)):
+            if _Primitive.is_primitive(ret[idx]):
+                ret[idx] = _Primitive.get_tensor(ret[idx], device)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:71]: I think my previous comment is just answered here, but I want to double check. This is wrapping kwargs with primitive types, right?

Use list comprehension to get better perf. List comprehension can "predict"the number of iterations and plan allocation ahead of time better than plain loop

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 6]
```diff
@@ -305,12 +334,17 @@ def _add_dynamic_shape(name, input):
...
-                input_names_require_grad.append(name)
+        if input is not None:
+            if (onnx_graph is None or name in onnx_graph_input_names):
+                if isinstance(input, torch.Tensor):
+                    if input.requires_grad:
+                        input_names_require_grad.append(name)
+                    dynamic_axes.update(_add_dynamic_shape(name, input))
+                    input_shape.append(list(input.size()))
+
+            # InputInfo should contain all the names irrespective of whether they are
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:345]: Double check this part. I remember using the fact that input_names only contains *used* inputs on my PRs regarding to kwargs (the last 2)

Review those PRs and make sure this change do not break anything that I might have failed in writing UTs for

File: orttraining/orttraining/python/training/ortmodule/_io.py [hunk 7]
```diff
@@ -305,12 +334,17 @@ def _add_dynamic_shape(name, input):
         return dynamic_axes
 
     def _add_input(name, input, onnx_graph, onnx_graph_input_names):
-        if input is not None and (onnx_graph is None or name in onnx_graph_input_names):
-            if input.requires_grad:
-                input_names_require_grad.append(name)
+        if input is not None:
+            if (onnx_graph is None or name in onnx_graph_input_names):
+                if isinstance(input, torch.Tensor):
+                    if input.requires_grad:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_io.py:340]: 4 levels of `if` give a hard time for the optimizer to properly parallelize or speed up the CPU pipeline. If you can optimize this somehow, that would be nice

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 8]
```diff
@@ -73,7 +73,7 @@ def forward(self, *inputs, **kwargs):
                                                       kwargs)
 
         # Reinitialize graph builder if the inputs or initializers requiring gradient have changed.
-        build_gradient_graph = build_gradient_graph or self._reinitialize_graph_builder(input_info)
+        build_gradient_graph = self._reinitialize_graph_builder(input_info) or build_gradient_graph
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:76]: Did this fix swap any subtle issue?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 9]
```diff
@@ -2252,3 +2252,72 @@ def run_step(expected, a, b, c, d, e, f, y, z):
                  a, b, c, d, e, f, y, z)
         run_step(a.item() + b.item() + c.item() + d.item() + e.item() + f.item() + y.item() + z.item(),
                  **{'a': a, 'b': b, 'c': c, 'd': d, 'e': e, 'f': f, 'y': y, 'z': z})
+
+@pytest.mark.parametrize("bool_argument", [True, False])
+@pytest.mark.parametrize("int_argument", [100, 100000, 100000000])
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2257]: try some negatives for int and float :)

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 10]
```diff
@@ -2252,3 +2252,72 @@ def run_step(expected, a, b, c, d, e, f, y, z):
...
+    class PrimitiveTypesInputNet(torch.nn.Module):
+        def __init__(self, input_size, hidden_size, num_classes):
+            super(PrimitiveTypesInputNet, self).__init__()
+
+            self.fc1 = torch.nn.Linear(input_size, hidden_size)
+            self.relu = torch.nn.ReLU()
+            self.fc2 = torch.nn.Linear(hidden_size, num_classes)
+
+        def forward(self, input1, bool_argument, int_argument, float_argument):
+            input1 = input1 + int_argument + float_argument
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2269]: assert types of bool_argument, int_argument and float_argument

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 11]
```diff
@@ -2252,3 +2252,72 @@ def run_step(expected, a, b, c, d, e, f, y, z):
                  a, b, c, d, e, f, y, z)
         run_step(a.item() + b.item() + c.item() + d.item() + e.item() + f.item() + y.item() + z.item(),
                  **{'a': a, 'b': b, 'c': c, 'd': d, 'e': e, 'f': f, 'y': y, 'z': z})
+
+@pytest.mark.parametrize("bool_argument", [True, False])
+@pytest.mark.parametrize("int_argument", [100, 100000, 100000000])
+@pytest.mark.parametrize("float_argument", [1.23, 11209123.12452, 12093702935.1249863])
+def test_primitive_inputs(bool_argument, int_argument, float_argument):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:2259]: assert types of bool_argument, int_argument and float_argument

**Review by thiagocrepaldi — COMMENTED** [review]
Nice job. Onecomment is that after *flattening* the input to pass it t the exporter, I was expecting that a *unflattening* would be needed before calling `FlattenedModule.forward`

Are we not unflatenning and exporting the onnx graph with tensors wrapping the primitive types instead? Is that kintentional?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 8 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py:76]: What I realized here is that if ```build_gradient_graph``` was ```True```, then ```self._reinitialize_graph_builder(input_info)``` would not be called at all.
So, this swap is needed to ensure that ```self._reinitialize_graph_builder(input_info)``` is called even if ```build_gradient_graph``` is ```True```.

File: orttraining/orttraining/python/training/ortmodule/_io.py (hunk 6 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_io.py:345]: This was actually a subtle bug we introduced earlier. Here is the illustration of the bug:

- Imagine we have a model with a boolean input:
```py
class PrimitiveTypesInputNet(torch.nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(PrimitiveTypesInputNet, self).__init__()

        self.fc1 = torch.nn.Linear(input_size, hidden_size)
        self.relu = torch.nn.ReLU()
        self.fc2 = torch.nn.Linear(hidden_size, num_classes)

    def forward(self, input1, bool_argument, int_argument, float_argument):
        input1 = input1 + int_argument + float_argument
        if bool_argument:
            out = self.fc1(input1)
            out = self.relu(out)
            out = self.fc2(out)
        else:
            out = self.fc1(input1)
            out = self.fc2(out)
            out = self.relu(out)
        return out
```
- On export, the bool input will be dropped in the onnx model.
- We would also have the ```bool_argument``` not a part of the ```input_info.names```. So, ```input_info.names``` would be ```[input1, int_argument, float_argument]```
- The inputs defined in the onnx model also would be ```[input1, int_argument, float_argument]```
- When the user provides inputs like ```(torch.tensor([1, 2, 3], False, 123, 1.234)``` to the forward method and since we only iterate over the onnx model inputs, we would see that the the the ```int_argument``` corresponds to the 2nd position in the input list, we would pick up the value of False, for the ```int_argument``` which is incorrect.

Basically, the ```input_info``` must contain all the non ```None``` inputs to the forward method, regardless of whether the input belongs to the onnx graph.

File: orttraining/orttraining/python/training/ortmodule/_io.py (hunk 7 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_io.py:340]: Unfortunately, I am not able to optimize this. However, I may be able to exit early and reduce the visibility of 4 checks.

**Comment by baijumeswani** [comment]
Unflattening is needed only when the ```args``` and ```kwargs``` need to to be extracted from a list of positional arguments.
We are still unflattening the input args into positional args and keyword kwargs. There is no need to reconvert the tensor wrapped primitive type, back to their native primitive form before executing the forward method (it would yield the same result).

File: orttraining/orttraining/python/training/ortmodule/_io.py (hunk 4 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_io.py:68]: Yes, this has been accounted for in the loop below.

File: orttraining/orttraining/python/training/ortmodule/_io.py (hunk 5 above)
Comment (baijumeswani) [reply orttraining/orttraining/python/training/ortmodule/_io.py:71]: I will try to use a list comprehension instead.

**Review by thiagocrepaldi — APPROVED** [review]
Great job

**Comment by baijumeswani** [comment]
Thanks. :)
